### PR TITLE
Firestore: implement equality semantics for public types

### LINKED ISSUE #6552: [Firestore] implement `__eq__` for public api types.
None of the public API types implement isEquals (__eq__)

### PR DESCRIPTION
**Pull request by tseaver**
Adds `__eq__` for the following types:

- `ArrayRemove`
- `ArrayUnion`
- `CollectionReference`
- `DocumentSnapshot`
- `ExistsOption`
- `LastUpdatedOption`
- `Query`

Closes #6552.

### CODE REVIEW

**Comment by tseaver** [comment]
@crwilcox Failed systests are all due to quota issues in the `watch` tests.

File: firestore/google/cloud/firestore_v1beta1/document.py [hunk 1]
```diff
@@ -547,6 +550,16 @@ def __init__(self, reference, data, exists, read_time, create_time, update_time)
         self.update_time = update_time
         """google.protobuf.timestamp_pb2.Timestamp: Document's last update."""
 
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return NotImplemented
+        return self._reference == other._reference and self._data == other._data
+
+    def __hash__(self):
```
[the remark this answers is not in the document]
Comment (tseaver) [reply firestore/google/cloud/firestore_v1beta1/document.py]: It is an interesting question, given that it is (theoretically) mutable.

[the remark this answers is not in the document]
Comment (mcdonc) [reply firestore/google/cloud/firestore_v1beta1/document.py]: I'm sure it's never mutated in the current codebase (and should never be), but I think it'd theoretically be possible to have two snapshots with the same read time and reference but different info.  Maybe just merge the listen-conformance branch into this branch temporarily and see if anything fails when the conformance tests run.  If not, it's probably fine.

[the remark this answers is not in the document]
Comment (tseaver) [reply firestore/google/cloud/firestore_v1beta1/document.py]: 2f1ea29

[the remark this answers is not in the document]
Comment (tseaver) [reply firestore/google/cloud/firestore_v1beta1/document.py]: @mcdonc Actually, given that the data is supposed to be immutable, I think the *correct* way to compute the snapshot hash is using the `updated_time` field, which is guaranteed to be set *by the server* at nanosecond precision (`read_time` is not set for snapshots retrieved via `DocumentReference.set`).

Having two objects hash to the same value but compare unequal is actually OK (versus having them compare equal but hash to different values:  see [Hynek's nice summary](https://hynek.me/articles/hashes-and-equality/)).

**Pull request merged**